File: anb/leetcode/array/238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
		vector<int> ans;
        int left=1, i, right=1;
		
		for(i=0;i<nums.size();i++) {
			ans.push_back(left);
			left *= nums[i];
		}
        for(i=nums.size()-1;i>=0;i--) {
            ans[i] *= right;
            right *= nums[i];
        }
		return ans;
    }
};

File: gunkim/leetcode/array/238.Product of Array Except Self.java
import java.util.Arrays;

public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(Arrays.toString(solution.productExceptSelf(new int[]{1,2,3,4})));
    }
}
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answers = new int[nums.length];

        int output = 1;
        for (int i = 0; i < nums.length; i++) {
            answers[i] = output;
            output *= nums[i];
        }
        output = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            answers[i] *= output;
            output *= nums[i];
        }
        return answers;
    }
}